Feature: patient registration report

jniles opened the pull request:
This pull request implements the patient registration report as a PDF report on the client.  It is opened by default via a "print" button on the patient registry page.  The report is available as JSON, HTML, and PDF using the renderers implemented by @sfount.

**Screenshots**
![html](https://cloud.githubusercontent.com/assets/896472/15963325/f3d0d0e6-2f08-11e6-9ba9-052c8cc17756.png)
_[Fig 1]: HTML Report_
![pdf](https://cloud.githubusercontent.com/assets/896472/15963326/f41ccadc-2f08-11e6-962a-a3c2e34bdcfd.png)
_[Fig 2]: PDF Report_

This is a very rough PR, and should not be considered final.  Here are some TODOs:
1. Filtering on the client should be additive, and use a key:value model, so that each filter takes in one parameter.  For example, `name:jniles` would be stored in the JSON `{ name : 'jniles' }`, and  be sent to the server as `?name=jniles`.
2. The Receipt Service is now overloaded with reports.  We need to cleanly separate Receipts as a subset of reports, and ensure that the modal and controllers are appropriately named.

These should be tackled in a subsequent PR.

Closes #470.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

jniles committed:
feat(patients): separate search() and find()

This commit refactors the patients search functionality in the following
ways:
 1. Expose a programmatic API via `find()` to the query engine.
 2. Make `patients/search/?` use `find()` internally.
 3. Rename `detail` to `detailed`.
 4. Remove `fields` support.  Now any query parameter will be used to
 filter the columns from the database.

This provides a much richer API for searching for patients on both the
client and the server.

jniles committed:
feat(patient registrations): implement pdf report

This commit refactors the server side to complete the PDF report
implementation.  Some notable improvements:
 1. Automatically template in the filters. This should be improved with
 varied colorschemes (IMA colors, for example).
 2. Adds `timestamp` helper to handlebars for more detailed time
 templates.
 3. Adds initial bare-bones clients structure for PDF printing modal.

jniles committed:
feat(registration report): implement clientside

This commit implements the client PDF report using the Receipt service
created by @sfount.  Because the client-side filters do not have a good
data model, printing with filtering has been left for later.

There are several prominent TODOs in this commit:
 1. Filtering on the client should be additive, and use a key:value
   model, so that each filter takes in one parameter.  For example,
   `name:jniles` would be stored in the JSON `{ name : 'jniles' }`, and
   be sent to the server as `?name=jniles`.
 2. The Receipt Service is now overloaded with reports.  We need to
 cleanly separate Receipts as a subset of reports, and ensure that the
 modal and controllers are appropriately named.

sfount commented on client/src/js/services/receipts/ReceiptModal.js:
```diff
@@ -44,7 +45,7 @@ function ReceiptModal(Modal, Receipts) {
     };
 
-    var invoiceRequest = Receipts.invoice(uuid, { render : options.renderer });
+    var invoiceRequest = Receipts.invoice(uuid, { renderer : options.renderer });
```
:+1:

sfount commented on client/src/js/services/receipts/ReceiptService.js:
```diff
@@ -18,39 +18,54 @@ function ReceiptService($http, util) {
-    if (options.render === renderers.PDF) { 
+
+    if (options.renderer === renderers.PDF) {
+      responseType = 'arraybuffer';
```
Just a note to say that in the future we could wrap this funcitonality in a utility method - there are a lot of crucial constants to get wrong here. 

First thing that comes to mind:

``` js
let params;
if (options.renderer === renderers.PDF) { 
  // returns 
  // { 
  //   params : options,
  //   responseType : 'arraybuffer'
  //  }
  params = utility.configurePDFParams(options);
}

$http.get(route, params);
```

What do you think?
@IMA-WorldHealth/local-contributors

jniles replied:
Sounds good to me.

sfount commented on server/controllers/medical/patients/index.js:
```diff
@@ -284,137 +286,82 @@ function hospitalNumberExists(req, res, next) {
-  })
-  .catch(next)
-  .done();
+ * @function logVisit
```
All of this was removed in https://github.com/IMA-WorldHealth/bhima-2.X/pull/477

It must have been reintroduced by accident in a rebase - can this be removed?

EDIT 
- That's anything to do with logVisit / visit

jniles replied:
Gotcha.  I was a bit confused in the changelog about what happened and favored preserving old code to removing it.  I'll remove it now.

sfount commented on server/controllers/medical/reports/patient.receipt.js:
```diff
@@ -10,36 +10,36 @@ supportedRender.html = require('../../../lib/renderers/html');
-  
-  var renderTarget = queryString.render || defaultRender;
+
+  var renderTarget = queryString.renderer || defaultRender;
```
:+1:

jniles committed:
fix(review): rm ancient logVisit functionality

This commit removes the logVisit functionality accidentally re-introduced
in a failed rebase.  The `logVisit` functionality has been superseded
by patient checkins.
